### PR TITLE
Initial support for FreeBSD kernel booting

### PR DESCRIPTION
**Pull request by duda-patryk**
Changes includes:
- FreeBSD kernel recognition
- Get proper entry point address
- Setup kernel trampoline

### CODE REVIEW

**Comment by nwhitehorn** [comment]
I worry some of this is a little too FreeBSD-specific. We also have in FreeBSD trunk support for vanilla kexec-lite now, but had to work through the following bugs to get there:
- kexec-lite's trampoline does not set r6 and r7 to zero (or to the ePAPR cookie), which is required by the standard
- Some of the asserts (the interpreter one, mostly, along with the ET_DYN/ET_EXEC check) are not appropriate and enforce more than required by ePAPR
- It doesn't use the entry point information for either ELFv1 or ELFv2 kernels, instead beginning execution at the first byte in the first loadable section. This patch addresses the last, but in a FreeBSD-specific block.

I think all of these could be fixed easily enough without requiring separate Linux and FreeBSD support code. For ELFv1/ELFv2 discrimination that works with both FreeBSD and Linux, skiboot already has appropriate code.

**Comment by cyrilbur-ibm** [comment]
Hi @semihalf-duda-patryk,
Thanks very much for the patch, it looks good. I do agree with Nathan that there is some FreeBSD specific stuff that I'm not convinced need to be.

It seems Nathan picked up on a few issues that I'm unware of, so thanks also to you Nathan.

I'm mostly unsure why a lot of the new code needs to be so FreeBSD specific.
For example, Linux needs correct values in r3, r4 and a zeroed r5. It looks like FreeBSD needs more parameters, it isn't clear to me why this requires a brand new trampoline. Surely we can come up with a single trampoline which suits both.

Thanks,

Cyril

**Comment by duda-patryk** [comment]
Thank you for the reply.

It turned out that I don't need another trampoline for FreeBSD. Now trampoline also sets r6 (to ePAPR cookie), and r7 to 0.
It doesn't need FreeBSD kernel recognition too.
It uses entry point information for obtaining entry point address.

If everything is OK and it meets your expectations, please merge this pull request. If not...comments are welcome!

File: kexec.c [hunk 1]
```diff
@@ -108,6 +108,49 @@ static void add_kexec_segment(char *type, void *buf, unsigned long bufsize,
 	kexec_segment_nr++;
 }
 
+static GElf_Shdr *getshdr(Elf *e, int idx, GElf_Shdr *shdr)
+{
+	if (gelf_getshdr(elf_getscn(e, idx), shdr) == NULL) {
+		fprintf(stderr, "load_kernel_freebsd: get section failed\n");
```
Comment (nwhitehorn) [inline kexec.c]: Should this still reference "load_kernel_freebsd"?

**Review by nwhitehorn — COMMENTED** [review]
This looks good to me and I think makes kexec-lite fully ePAPR-compliant to boot.

File: kexec.c (hunk 1 above)
Comment (duda-patryk) [reply kexec.c]: Of course it shouldn't. Fixed.

**Comment by cyrilbur-ibm** [comment]
Yeah, looks good to me too. I tested by using it in petitboot and works fine. Not really sure if the review process on github works like this but whatever:

Reviewed-by: Cyril Bur <cyrilbur@gmail.com>

**Comment by antonblanchard** [comment]
Thanks everyone, merged.

**Pull request merged**